View a past order - closes 25

Laszlo-JFLMTCO opened the pull request:
@Dpalazzari @drod1000 @njgheorghita 
View a past order implementation. This branch has the `40-order-items` branch from @drod1000 already merged. I added four user stories as the `show` page content should be dynamic based on the `order.status`.
With this branch all tests pass.

Before further work, we should have a decision on whether we really rename `order_items` to `order_outings` as all further user stories will rely on this detail, so we can contain the amount of rewrite now. Please let me know what you think.

Side note: there might be inconsistencies through the user stories, but originally our `order` model did not have `completed` enum, but had `shipped` instead. For now I left `shipped` in the enum array and added `completed` until we can confirm the list of the proper `order.status` values.

drod1000 committed:
Adds tests for OrderItem

drod1000 committed:
Adds OrderItem model and table

drod1000 committed:
Adds relationships to order items

drod1000 committed:
Adds subtotal test for order_item

drod1000 committed:
Adds subtotal for order items

drod1000 committed:
Adds order to order-items 1-many

drod1000 committed:
Adds total method to order

Laszlo-JFLMTCO committed:
WIP, Create test for user story when user views a past order

Laszlo-JFLMTCO committed:
Merge branch '40-order-items' of https://github.com/drod1000/lobbify into 40-order-items
Creating local copy of 40-order-items branch to allow user ticket 25 development.

Laszlo-JFLMTCO committed:
Add feature a user can view a past order feature, corner cases are not tested yet

Laszlo-JFLMTCO committed:
Expand test coverage for all order status ordered, paid, cancelled and completed

drod1000 committed:
Renames order-items to order-outings

drod1000 committed:
Adds join table between order and outing

drod1000 committed:
Updates total test for order

drod1000 committed:
Adds currency method to order-outing and order

drod1000 commented on app/views/orders/show.html.erb:
```diff
@@ -0,0 +1,28 @@
+      </tr>
+    <% end %>
+  </table>
+</div>
```
I might be missing it because it looks like the test is passing but but can you point out where the order total is on this view? I'm only seeing the item, quantity, and the subtotals.

Dpalazzari replied:
I think the subtotal is the total? @Laszlo-JFLMTCO @drod1000

Laszlo-JFLMTCO replied:
Good catch @drod1000 and @Dpalazzari . I will have to take a second look why the test is passing without `show.html.erb` view not printing the order total. I will look into that and get back to you guys! @njgheorghita

Laszlo-JFLMTCO replied:
@drod1000 @Dpalazzari @njgheorghita Quick update: again very good catch of missing the total from the page. You guys just highlighted a weakness in the test cases. Since the order only had a single outing, even if its quantity was more than one, the `subtotal` was essentially the same as `total` and that is why the test passed. I am working on making the test cases stronger to ensure we only pass if total is display and the correct `order.total` is displayed. Thank you again for highlighting this!!!

drod1000 commented:
@Laszlo-JFLMTCO Great work. I'm only seeing completed in the user stories, where do you see shipped? As for the tests I was wondering why you created 4 orders when we're only visiting the first order in the tests? I believe you guys were right about renaming the table to order-outings. I made those changes and updated that PR so that we have the correct name once it gets merged.

Dpalazzari commented on app/controllers/orders_controller.rb:
```diff
@@ -2,4 +2,12 @@ class OrdersController < ApplicationController
+
+  def show
+    @order = Order.find(params[:id])
+    if !current_user.orders.include?(@order)
```
@Laszlo-JFLMTCO Can you explain this line to me?

Laszlo-JFLMTCO replied:
@Dpalazzari This is to make sure we only show orders belong to the logged in user (which is `current_user`). With our current implementation without this check, one user could log in, and enter the path of `/orders/:id` where the `:id` would refer to an order associated with another user. We are checking here that the order id we received through `params[:id]` is part of the collection of orders of the user that is logged in `current_user.orders`. FYI @njgheorghita @drod1000  Please let me know in case further explanation is necessary.

drod1000 committed:
Removes self from orderouting subtotal method

Laszlo-JFLMTCO committed:
Merge branch '40-order-items' of https://github.com/drod1000/lobbify into 40-order-items
Pulling latest updates from GitHub repository.

Laszlo-JFLMTCO committed:
Schema file updated after migration

Laszlo-JFLMTCO committed:
Merge branch '40-order-items' into 25-view-a-past-order
Local merge in preparation to update 25-view-a-past-order PR up on GitHub

Laszlo-JFLMTCO committed:
Improve test cases to cover order.total and corner cases as user not logged in or logged in but wants to check other users order

Laszlo-JFLMTCO commented:
@drod1000 @Dpalazzari @njgheorghita I added much improvement to the test cases, updated the order `show.html.erb` to accomodate the `order.total`. Merged with the latest branch updated from @drod1000 
All tests pass. Please let me know if you think I missed any other key detail.

njgheorghita commented on app/models/order.rb:
```diff
@@ -1,5 +1,19 @@
 class Order < ApplicationRecord
+  include ActionView::Helpers::NumberHelper
```
same issue as before? sorry for being a nitpicker

njgheorghita commented:
looks good to me, asides from that one redundant line but we can deal with that in refactoring, i'm going to go ahead and merge
